### DISCUSSION TITLE
AWS Cloud9 or Code Editor Environment "[Error 401] access denied, invalid bearer token" when authenticating

### DISCUSSION THREAD

**Discussion by b1tf0rbyt3s** [CrowdStrike/falconpy#1132] [Q&A]
Was attempting to utilize falconpy from within our AWS coding environment.  Tried creating a Cloud9 environment with both the Amazon Linux and Ubuntu 22 EC2 underlying it.  When I try to run any calls I get a [Error 401] access denied, invalid bearer token.  Cloud9 comes default with Python 3.9.16 when using Amazon Linux OS and 3.10 when using Ubuntu 22.  I even tried upgrading to 3.12 to see if there was any difference.  I also tried using Code Editor from within AWS SageMaker and it defaults to python 3.10 - get exact same error.  

What is puzzling is that the same exact code works fine when running within my local IDE on my laptop, utilizing same python versions.  Makes me think it is something within the AWS environment that is not allowing authentication but not sure how to pinpoint the exact issue as it only gives me a "access denied, invalid bearer token" error.  

Sample test code to demonstrate how I am attempting authenticate (This code works perfectly in local IDE but gives the error mentioned above in code editing environments within AWS):

```python
import os
from falconpy import APIHarnessV2

falcon = APIHarnessV2(client_id=os.getenv("FALCON_CLIENT_ID"),
                      client_secret=os.getenv("FALCON_CLIENT_SECRET"))

SEARCH_FILTER = "hostname:'HOSTNAME'"
response = falcon.command("QueryDevicesByFilter",
                          limit=1,
                          filter=SEARCH_FILTER
                          )
if response["status_code"] == 200:
    hosts_found = response["body"]["resources"]
    print(f'Host ID Found: {hosts_found}')
else:
    # Retrieve the details of the error response
    error_detail = response["body"]["errors"]
    for error in error_detail:
        # Display the API error detail
        error_code = error["code"]
        error_message = error["message"]
        print(f"[Error {error_code}] {error_message}")
```
On Cloud9 or AWS Code Editor I get:
[Error 401] access denied, invalid bearer token

On local IDE, I get 200 and the Host ID every time.

I also tested same code on personal AWS account with full admin rights.  Same result.

Any ideas as to what could be restricting from within AWS environment or best way to narrow down what might be the issue?

**Comment by jshcodes** [comment]
The example above appears to leverage [Environment Authentication](https://www.falconpy.io/Usage/Authenticating-to-the-API.html#environment-authentication) (manually).  If you swap to [Direct Authentication](https://www.falconpy.io/Usage/Authenticating-to-the-API.html#direct-authentication), do you get a valid bearer token?

**Reply by b1tf0rbyt3s** [reply]
What would Direct Auth look like for the above example.  Not sure I fully understand the difference and want to test it out.

**Reply by jshcodes** [reply]
The example above is pulling the API credentials from the environment.

```python
falcon = APIHarnessV2(client_id=os.getenv("FALCON_CLIENT_ID"),
                      client_secret=os.getenv("FALCON_CLIENT_SECRET"))
```
This recreates [Environment Authentication](https://www.falconpy.io/Usage/Authenticating-to-the-API.html#environment-authentication) functionality and for most scenarios is unnecessary in newer version of FalconPy. Instead, we can create the instance of the Uber class like this.

```python
falcon = APIHarnessV2()
```

I'm curious to see if the API credentials are actually present within the Cloud9 environment. One way to confirm that would be to create an instance of the Uber class using [Direct Authentication](https://www.falconpy.io/Usage/Authenticating-to-the-API.html#direct-authentication), which provides the API client ID and secret as strings via keyword arguments.

```python
# Do not hard code credentials as demonstrated below in production environments.
# We're doing this now as a quick test to confirm the environment is sane.
falcon = APIHarnessV2(client_id="API CLIENT ID GOES HERE", client_secret="API CLIENT SECRET GOES HERE")

results = falcon.command("QueryDevicesByFilterScroll", limit=1)
print(results)
```
If we receive a valid API response, then more than likely we need to [add the API keys as environment variables to the Cloud9 environment](https://docs.aws.amazon.com/cloud9/latest/user-guide/env-vars.html) to leverage [Environment Authentication](https://www.falconpy.io/Usage/Authenticating-to-the-API.html#environment-authentication).

> Another way to confirm this would be to call `env` from the Cloud9 terminal.

**Reply by b1tf0rbyt3s** [reply]
Ok, that makes sense.  Thanks.  I ran the following tests in brand new Cloud9 environments:

### Test 1 - Cloud9 SSM Environment (AWS Systems Manager Connection) configured with a Amazon Linux 2023 EC2 -> Python 3.9.16

- Installed FalconPy 1.4.1
```bash
python3 -m pip install crowdstrike-falconpy
Defaulting to user installation because normal site-packages is not writeable
Collecting crowdstrike-falconpy
  Downloading crowdstrike_falconpy-1.4.1-py3-none-any.whl (638 kB)
     |████████████████████████████████| 638 kB 5.1 MB/s            
Requirement already satisfied: urllib3 in /usr/lib/python3.9/site-packages (from crowdstrike-falconpy) (1.25.10)
Requirement already satisfied: requests in /usr/lib/python3.9/site-packages (from crowdstrike-falconpy) (2.25.1)
Requirement already satisfied: chardet<5,>=3.0.2 in /usr/lib/python3.9/site-packages (from requests->crowdstrike-falconpy) (4.0.0)
Requirement already satisfied: idna<3,>=2.5 in /usr/lib/python3.9/site-packages (from requests->crowdstrike-falconpy) (2.10)
Installing collected packages: crowdstrike-falconpy
Successfully installed crowdstrike-falconpy-1.4.1
```
Verified there were no env variables for Falcon
```bash
~/environment $ env |grep FALCON
```
- Tested the following code:
```python
from falconpy import APIHarnessV2

falcon = APIHarnessV2(client_id="<REDACTED>", client_secret="<REDACTED>")

results = falcon.command("QueryDevicesByFilterScroll", limit=1)
print(results)
```
- Got the following output:
```bash
{'status_code': 401, 'headers': {'Server': 'nginx', 'Date': 'Fri, 22 Mar 2024 18:47:04 GMT', 'Content-Type': 'application/json', 'Content-Length': '231', 
'Connection': 'keep-alive', 'X-Content-Type-Options': 'nosniff', 'X-Cs-Traceid': '66689551-3112-48e6-8ae3-e967e044d766', 'X-Ratelimit-Limit': '15', 
'X-Ratelimit-Remaining': '14', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'}, 'body': {'meta': {'query_time': 1.94e-07, 
'powered_by': 'crowdstrike-api-gateway', 'trace_id': '66689551-3112-48e6-8ae3-e967e044d766'}, 'errors': [{'code': 401, 'message': 'access denied, invalid bearer token'}]}}
```
- Added the env variables
```bash
~/environment $ export FALCON_CLIENT_ID=<REDACTED>
~/environment $ export FALCON_CLIENT_SECRET=<REDACTED>
~/environment $ env |grep FALCON
FALCON_CLIENT_SECRET=<REDACTED>
FALCON_CLIENT_ID=<REDACTED>
```
- Tested Direct with the following code
```python
from falconpy import APIHarnessV2

falcon = APIHarnessV2()

results = falcon.command("QueryDevicesByFilterScroll", limit=1)
print(results)
```
Got same output as above: 
**'errors': [{'code': 401, 'message': 'access denied, invalid bearer token'}]**

- Falcon API Audit Trail for that Client ID (The IP is whitelisted within Falcon CID)
```bash
2024-03-22 18:47:04	API request logged	POST /oauth2/token	Request: POST /oauth2/token, Query: , Status: 403, Scope: , IP: <REDACTED>, User-Agent: crowdstrike-falconpy/1.4.1
```
(Got same log entry for all tests above and below - Status 403)

### Test 2 - Conducted same tests with new Cloud9 environment with same code as above and got same results as above:
Cloud9 SSM Environment (AWS Systems Manager Connection) configured with a **Ubuntu 22.04** EC2 -> **Python 3.10.12** 

- Install of FalconPy 1.4.1
```bash
python3 -m pip install crowdstrike-falconpy
Defaulting to user installation because normal site-packages is not writeable
Collecting crowdstrike-falconpy
  Downloading crowdstrike_falconpy-1.4.1-py3-none-any.whl.metadata (20 kB)
Requirement already satisfied: requests in /usr/lib/python3/dist-packages (from crowdstrike-falconpy) (2.25.1)
Requirement already satisfied: urllib3 in /usr/lib/python3/dist-packages (from crowdstrike-falconpy) (1.26.5)
Downloading crowdstrike_falconpy-1.4.1-py3-none-any.whl (638 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 638.5/638.5 kB 10.8 MB/s eta 0:00:00
Installing collected packages: crowdstrike-falconpy
Successfully installed crowdstrike-falconpy-1.4.1
```
### Test 3 - I reran same code in local environment and actually got same 401 error for direct auth (not sure why?), but successful 200 API request when specifying the client ID/secret:
```python
falcon = APIHarnessV2(client_id="<REDACTED>", client_secret="<REDACTED>")
```
Output:
```bash
status_code': 200, 'headers': {'Server': 'nginx', 'Date': 'Fri, 22 Mar 2024 19:02:43 GMT' ...
```
### Test 4 - Used different AWS account - my personal account (to rule out that it was restriction based on my IAM credentials) - same errors as above
**'errors': [{'code': 401, 'message': 'access denied, invalid bearer token'}]**

**Reply by jshcodes** [reply]
Does your API client have the __Hosts__ _READ_ scope assigned to it?

**Reply by b1tf0rbyt3s** [reply]
It does.  The enabled scopes are:

- Hosts: Read
- Real time response (Admin): Write
- Real time response audit: Read
- Real time response: Read + Write 

Currently running several batch jobs with that API client ID for putting a custom forensics collector, running it, doing batch Gets to collect the result zip archive - but utilizing PSFalcon manually - and locally.  I want to eventually move to Python as part of the automated pipeline where we upload to S3 and have Lambda functions (thus why it would be easier to work within AWS IDE) that preprocess the collected information, and enriches it, before sending it to our log aggregator for further analysis.

**Reply by jshcodes** [reply]
It's weird we're getting a 401 back when trying the following code (if our client ID and secret are valid):

```python
from falconpy import APIHarnessV2

falcon = APIHarnessV2(client_id="<REDACTED>", client_secret="<REDACTED>")

results = falcon.command("QueryDevicesByFilterScroll", limit=1)
print(results)
```

Can we take this code and execute it outside of Cloud9?  When doing so, let's turn on debugging just so we see the full interaction. (Keys will be redacted still.)

```python
import logging
from falconpy import APIHarnessV2

logging.basicConfig(level=logging.DEBUG)
falcon = APIHarnessV2(client_id="<REDACTED>", client_secret="<REDACTED>", debug=True)

results = falcon.command("QueryDevicesByFilterScroll", limit=1)
print(results)
```

> Another thing we should check: Does your tenant have IP allowlist management enabled?

**Reply by b1tf0rbyt3s** [reply]
Here is the output from same CID when I run via local IDE.  In terms of IP Allowlist - I will usually add the IP addresses to the allow list in Falcon - in this case I added the IP addresses that are logged in Falcon for Amazon and my own IP Address as well (for running locally).  

```bash
DEBUG:falconpy._auth_object._falcon_interface:CREATED: APIHarnessV2 interface class
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: Base URL set to https://api.crowdstrike.com
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: SSL verification is set to True
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: Timeout set to None seconds
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: Proxy dictionary: None
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: User-Agent string set to: None
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: Token renewal window set to 120 seconds
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: Maximum number of records to log: 100
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: Log sanitization is enabled
DEBUG:falconpy._auth_object._falcon_interface:CONFIG: Pythonic responses are disabled
DEBUG:falconpy._auth_object._falcon_interface:OPERATION: oauth2AccessToken
DEBUG:falconpy._auth_object._falcon_interface:ENDPOINT: https://api.crowdstrike.com/oauth2/token (POST)
DEBUG:falconpy._auth_object._falcon_interface:HEADERS: {'User-Agent': 'crowdstrike-falconpy/1.4.1', 'CrowdStrike-SDK': 'crowdstrike-falconpy/1.4.1'}
DEBUG:falconpy._auth_object._falcon_interface:PARAMETERS: None
DEBUG:falconpy._auth_object._falcon_interface:BODY: None
DEBUG:falconpy._auth_object._falcon_interface:DATA: {'client_id': 'REDACTED', 'client_secret': 'REDACTED'}
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.crowdstrike.com:443
DEBUG:urllib3.connectionpool:https://api.crowdstrike.com:443 "POST /oauth2/token HTTP/1.1" 308 0
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.us-2.crowdstrike.com:443
DEBUG:urllib3.connectionpool:https://api.us-2.crowdstrike.com:443 "POST /oauth2/token HTTP/1.1" 201 1251
DEBUG:falconpy._auth_object._falcon_interface:RECEIVED: Content returned in application/json format
DEBUG:falconpy._auth_object._falcon_interface:STATUS CODE: 201
DEBUG:falconpy._auth_object._falcon_interface:RESULT: {'status_code': 201, 'headers': {'Server': 'nginx', 'Date': 'Mon, 01 Apr 2024 16:00:32 GMT', 'Content-Type': 'application/json', 'Content-Length': '1251', 'Connection': 'keep-alive', 'X-Cs-Region': 'us-2', 'X-Cs-Traceid': '6af46408-f3f0-4907-9d35-77d540f76bcf', 'X-Ratelimit-Limit': '300', 'X-Ratelimit-Remaining': '298', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains'}, 'body': {'access_token': 'REDACTED', 'expires_in': 1799, 'token_type': 'bearer'}}
DEBUG:falconpy._auth_object._falcon_interface:OPERATION: QueryDevicesByFilterScroll
DEBUG:falconpy._auth_object._falcon_interface:ENDPOINT: https://api.us-2.crowdstrike.com/devices/queries/devices-scroll/v1 (GET)
DEBUG:falconpy._auth_object._falcon_interface:HEADERS: {'Authorization': 'Bearer REDACTED', 'User-Agent': 'crowdstrike-falconpy/1.4.1', 'CrowdStrike-SDK': 'crowdstrike-falconpy/1.4.1'}
DEBUG:falconpy._auth_object._falcon_interface:PARAMETERS: {'limit': 1}
DEBUG:falconpy._auth_object._falcon_interface:BODY: {}
DEBUG:falconpy._auth_object._falcon_interface:DATA: {}
DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): api.us-2.crowdstrike.com:443
DEBUG:urllib3.connectionpool:https://api.us-2.crowdstrike.com:443 "GET /devices/queries/devices-scroll/v1?limit=1 HTTP/1.1" 200 420
DEBUG:falconpy._auth_object._falcon_interface:RECEIVED: Content returned in application/json format
DEBUG:falconpy._auth_object._falcon_interface:STATUS CODE: 200
DEBUG:falconpy._auth_object._falcon_interface:RESULT: {'status_code': 200, 'headers': {'Server': 'nginx', 'Date': 'Mon, 01 Apr 2024 16:00:33 GMT', 'Content-Type': 'application/json', 'Content-Length': '420', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains, max-age=31536000; includeSubDomains', 'X-Cs-Region': 'us-2', 'X-Cs-Traceid': 'c88930c3-920d-4199-990a-120ae16221a7', 'X-Ratelimit-Limit': '6000', 'X-Ratelimit-Remaining': '5999'}, 'body': {'meta': {'query_time': 0.056376144, 'pagination': {'total': 14463, 'offset': 'FGluY2x1ZGVfY29udGV4dF91dWlkDnF1ZXJ5VGhlbkZldGNoAhYwTlZjNlhvaVM2LWFUYVp5NFBrMGh3AAAAACtZefAWMTBqYllacVpTd2lCZ21NVFNfQ1hyZxZkcjVSbG90V1Q4bXlXdUNDYVJiV1BnAAAAADIhV4EWV0RkSGVoNklUanVCdVZGMjNmbTNrUQ==', 'expires_at': 1711987353204342414}, 'powered_by': 'device-api', 'trace_id': 'c88930c3-920d-4199-990a-120ae16221a7'}, 'resources': ['4cfd4e7ed50f472b9aad5c0bedc839bf'], 'errors': []}}

{'status_code': 200, 'headers': {'Server': 'nginx', 'Date': 'Mon, 01 Apr 2024 16:00:33 GMT', 'Content-Type': 'application/json', 'Content-Length': '420', 'Connection': 'keep-alive', 'Content-Encoding': 'gzip', 'Strict-Transport-Security': 'max-age=15724800; includeSubDomains, max-age=31536000; includeSubDomains', 'X-Cs-Region': 'us-2', 'X-Cs-Traceid': 'c88930c3-920d-4199-990a-120ae16221a7', 'X-Ratelimit-Limit': '6000', 'X-Ratelimit-Remaining': '5999'}, 'body': {'meta': {'query_time': 0.056376144, 'pagination': {'total': 14463, 'offset': 'FGluY2x1ZGVfY29udGV4dF91dWlkDnF1ZXJ5VGhlbkZldGNoAhYwTlZjNlhvaVM2LWFUYVp5NFBrMGh3AAAAACtZefAWMTBqYllacVpTd2lCZ21NVFNfQ1hyZxZkcjVSbG90V1Q4bXlXdUNDYVJiV1BnAAAAADIhV4EWV0RkSGVoNklUanVCdVZGMjNmbTNrUQ==', 'expires_at': 1711987353204342414}, 'powered_by': 'device-api', 'trace_id': 'c88930c3-920d-4199-990a-120ae16221a7'}, 'resources': ['4cfd4e7ed50f472b9aad5c0bedc839bf'], 'errors': []}}

```

**Reply by jshcodes** [reply]
This one is pretty interesting. We've eliminated most of the low hanging fruit at this point. 🤔

In the Cloud9 environment, we can tweak the code slightly to force an authentication (the Uber Class doesn't authenticate until the first request is made). This will allow us to confirm a token is being generated.

> [!IMPORTANT]
> __Do not post the response of the following code to our GitHub discussion.__

```python
import logging
from falconpy import APIHarnessV2
logging.basicConfig(level=logging.DEBUG)

uber = APIHarnessV2(client_id="<REDACTED>", client_secret="<REDACTED>", debug=True)
uber.authenticate()
print(uber.token)
```

I'll setup a Cloud9 environment today and try this out to see if I can recreate the issue.

**Reply by b1tf0rbyt3s** [reply]
Tried running the above in Cloud9.  I get **None** for Token and is just a 403: {'code': 403, 'message': 'access denied, authorization failed'}.  

What is interesting is that for the code above where we tested the direct client id/secret it is a 403 in the debug output and then a 401 in same output at the end with the stdout of _print(results)_ - see below.   I don't know if that means anything? Would it be an issue with **urllib3**?

```bash

...
DEBUG:urllib3.connectionpool:https://api.crowdstrike.com:443 "POST /oauth2/token HTTP/1.1" 403 231
DEBUG:falconpy._auth_object._falcon_interface:RECEIVED: Content returned in application/json format
ERROR:falconpy._auth_object._falcon_interface:ERROR: access denied, authorization failed
DEBUG:falconpy._auth_object._falcon_interface:STATUS CODE: 403
...
...'errors': [{'code': 401, 'message': 'access denied, invalid bearer token'}]}}



```

**Reply by jshcodes** [reply]
Our authentication request is definitely failing, and we're pretty sure about the keys since we're using them locally.  If we call `curl -s ipinfo.io/ip` from the Cloud9 terminal, does it match the IP added to the Allowlist?

Both status codes are part of the API response. The 403 is from the `status_code` key, whereas the `errors` branch is a child branch of `body`.  That __*is*__ interesting. I will try and recreate this response in my testing.
